### ISSUE TITLE
Improve Kubernetes resources validation based on openAPIV3Schema

### ISSUE THREAD

**Issue by weisdd** [traefik/traefik#8338]
### Welcome!

- [X] Yes, I've searched similar issues on [GitHub](https://github.com/traefik/traefik/issues) and didn't find any.
- [X] Yes, I've searched similar issues on the [Traefik community forum](https://community.containo.us) and didn't find any.

## Missing functionality

Traefik currently doesn’t offer any linting functionality, so the only way to check whether your configuration is correct is just to apply the desired configuration and then follow traefik’s logs.

E.g., in case of errors in a middleware, you might end up with broken IngressRoutes:

```
  ipWhiteList:
    sourceRange:
      - 10.0.0.s
```

What's also a bit pity here is that the Middleware is not validated by Traefik until you refer to it in an IngressRoute.

Currently, Traefik has no ValidationWebhook (useful for deployments) or linter (useful for pipelines).

I've decided to make a PoC for linter. It'd be great if someone familiar with the traefik codebase can take over it and make it a production-ready solution :) I can certainly help with it, though my experience with Go is not that deep yet. 

## PoC

When I looked at the projects code, I found out that unit tests already contain mocked k8s client that can load manifests from files and let you play with the resources later on. Unfortunately, most of the methods I wanted to use are not exported and/or not available for calls outside tests, thus I needed to fork the project to rename some of the files and add a few methods.

* changes: https://github.com/weisdd/traefik/compare/f251394..71160da
* compiled binaries: https://github.com/weisdd/traefik/releases/tag/tralint-0.1.0

A few notes:

* Since linter should be able to work with any kubernetes resources, I disabled panicking when it encounters so called unknown runtime object;
* Since, again, it was only a PoC, all changes were made only to make my concept work, so they're likely to affect the traefik binary itself.
* In real cluster, Endpoints get populated automatically. With fake k8s client, I think it makes sense to generate them on the fly, so I added a method for that.
* The linter itself is quite rough/edgy at the moment, so, please, have low expectations :)

### Current limitations

* Normally, Traefik adds all errors to a special counter, so I can generate a non-zero exit code based on that. Though, there are still some errors that are not counted, so cannot fully rely on the exit code now. If you see `ERRO` in log, it's a more reliable indicator. Unfortunately, it writes data to stdout, so you cannot rely on catching stderr either.
* It cannot work with List objects, so you need to split them into a bunch of documents (e.g. with the help of jq: `kubectl get middleware -A -o yaml | yq eval '.items.[] | splitDoc' - > fixtures/middlewares.yaml`). - Limitation of the original package. This one can be fixed.
* For some "errors", tralint will print all things it's unhappy about. For others, like when important part of configuration is missing and traefik cannot build it's runtime config, tralint will print just that critical error. Once it's resolved, you'll have to run in again.
* Currently, HTTP entrypoints are hardcoded: `web`, `websecure` - should be enough for now;
* Manifests have to be stored in `./fixtures/` as in the original test code that I reused;
* Like with the original Traefik, a middleware is checked only if you have a manifest with an `IngressRoute` where you refer to that middleware;
* Since the tool builds up configuration for traefik runtime, you need to have manifests for all related "kinds" (`IngressRoute`, `Middleware`, `Service`). In real traefik setup, those are also needed. Ideally, we should be able to have some linting options that can say if individual resources have incorrect syntax.

### Usage

Due to helm and tralint limitations, we’ll need to first install [yq](https://github.com/mikefarah/yq/releases/latest).

tralint currently expects you to store all manifests in `fixtures/` (as the original test code that I reused), so we need to create it:

```
mkdir fixtures
```

Specify a namespace where you plan to deploy the chart:

```
export RELEASE_NAMESPACE=test2
````

Download a copy of all middlewares from our cluster, so you can refer to them from IngressRoutes. Since kubectl will return `kind: List`, we need to use yq to split the result into a set of YAML documents:

```
kubectl get middleware -A -o yaml | yq eval '.items.[] | splitDoc' - > fixtures/middlewares.yaml
```

Download a copy of a TLS certificate from the cluster. If you happen to use a wildcard certificate that gets copied to every namespace, you can save only that one.
In the example below, we save all certs from `traefik` namespace and replace `.metadata.namespace` with the namespace that's actually going to be used for a helm release (needed for IngressRoute validation):

```
kubectl get secrets --field-selector type=kubernetes.io/tls -n traefik -o yaml | yq eval '.items.[] | splitDoc' - | yq eval-all '.metadata.namespace = env(RELEASE_NAMESPACE)' - > fixtures/certs.yaml
```

Save all manifests from helm chart to a file through the template subcommand.
As an extra step, we need to add `.metadata.namespace` to every resulting manifest, because helm doesn't do that (and there are no plans to support this feature from their side):

```
helm template . | yq eval-all '.metadata.namespace = env(RELEASE_NAMESPACE)' - > fixtures/chart-manifests.yaml
```

Now, we can lint the manifests:

```
$ tralint
INFO[0000] Skipped *v1.ConfigMap as it's irrelevant for Traefik setup 
ERRO[0000] cannot parse CIDR whitelist [127.0.0.1s]: parsing CIDR trusted IPs <nil>: invalid CIDR address: 127.0.0.1s  entryPointName=web routerName=traefik-test-ingressroute-d1dbb6cdb7259555612d
INFO[0000] Processed 1 ingressRoutes, 1 middlewares, and some other resources 
FATA[0000] Found 2 errors
```

**Comment by mpl** [comment]
Hello. the following question is by no means a commitment in one way or the other, but are you asking:

1) that tralint becomes a standalone project maintained under the umbrella of traefik org?
2) or that what tralint does becomes part of the reload configuration phase of traefik proxy, everytime a kubernetes provider event is received?
3) something else?

**Comment by weisdd** [comment]
@mpl Since data types for API resources are generated automatically, I think a linter should be a part of traefik itself (not necessarily a separate binary), so you can always be sure that the linter plays by the same rules as traefik (=> not fails on the same set of manifests).
Having that, someone could run it locally against a full set of manifests to make sure configuration is correct (all dependent objects exist, syntax is correct, middlewares don't have circular dependencies, and so on). - That would be useful in pipelines / IDEs.

In the bright and distant Future, it'd be great to have a webhook that can validate configuration. It can reuse some of linting functionality, though it looks like a more complex thing to implement, because it needs to consider full set of resources, and, thus, sequencing of applying manifests matters.

I think that it'd make a lot of sense to make mock client and its methods public, so people can start experimenting more with it (using like a library) and contribute back.
Also, it'd be great to see more errors being added to traefik error counters, so it becomes easier to generate a correct exit code.
Independent checks for middlewares could be useful as well.

**Comment by rgstephens** [comment]
I agree that a linter or validate would be great. I just ran into [#7802](https://github.com/traefik/traefik/issues/7802) where a new invalid `IngressRoute` caused subsequent route changes to fail loading.

**Comment by mpl** [comment]
@weisdd Given that Kubernetes can already do some validation through the openAPIV3Schema constraints (see all the .`yaml` files starting wih `traefik.containo.us_`) in `./docs/content/reference/dynamic-configuration`), don't you think we could simply improve and expand these constraints to achieve what you want to do?
Is there anything in the "linting" that you're proposing that does not seem doable that way?

Of course it would still be up to the user to apply these constraint definitions along with all the other traefik CRDs they apply, but that's nothing new.

It seems to us that it would be preferable if all the validation could be done that way.

**Comment by weisdd** [comment]
@mpl I think it makes a lot of sense to let Kubernetes do the validation. It won't cover 100% of cases (like something very specific to Traefik's internals like circular dependencies - something that I saw in the code), though it's already good enough, and should be relatively easy to implement/extend and support rather than custom CLIs.
E.g. I've recently bumped in the case where I used named port instead of a port number in an IngressRoute, and it also broke watch logic like in @rgstephens case. Can easily be prevented by k8s validation.
